# Module 10 Application

## Challenge: Crypto Clustering

In this Challenge, you’ll combine your financial Python programming skills with the new unsupervised learning skills that you acquired in this module.

The CSV file provided for this challenge contains price change data of cryptocurrencies in different periods.

The steps for this challenge are broken out into the following sections:

* Import the Data (provided in the starter code)
* Prepare the Data (provided in the starter code)
* Find the Best Value for `k` Using the Original Data
* Cluster Cryptocurrencies with K-means Using the Original Data
* Optimize Clusters with Principal Component Analysis
* Find the Best Value for `k` Using the PCA Data
* Cluster the Cryptocurrencies with K-means Using the PCA Data
* Visualize and Compare the Results

### Import the Data

This section imports the data into a new DataFrame. It follows these steps:

1. Read  the “crypto_market_data.csv” file from the Resources folder into a DataFrame, and use `index_col="coin_id"` to set the cryptocurrency name as the index. Review the DataFrame.

2. Generate the summary statistics, and use HvPlot to visualize your data to observe what your DataFrame contains.


> **Rewind:** The [Pandas`describe()`function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) generates summary statistics for a DataFrame. 

In [126]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [127]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    Path("Resources/crypto_market_data.csv"),
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [128]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [129]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

This section prepares the data before running the K-Means algorithm. It follows these steps:

1. Use the `StandardScaler` module from scikit-learn to normalize the CSV file data. This will require you to utilize the `fit_transform` function.

2. Create a DataFrame that contains the scaled data. Be sure to set the `coin_id` index from the original DataFrame as the index for the new DataFrame. Review the resulting DataFrame.


In [130]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data)

In [131]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns
)

# Copy the crypto names from the original data
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data

In this section, you will use the elbow method to find the best value for `k`.

1. Code the elbow method algorithm to find the best value for `k`. Use a range from 1 to 11. 

2. Plot a line chart with all the inertia values computed with the different values of `k` to visually identify the optimal value for `k`.

3. Answer the following question: What is the best value for `k`?

In [132]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
# YOUR CODE HERE!

k = list(range(1,11))

In [133]:
# Create an empy list to store the inertia values
# YOUR CODE HERE!

inertia = []

In [134]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
# YOUR CODE HERE!



for i in k: 
    model = KMeans(n_clusters=i, random_state=1)
    model.fit(df_market_data_scaled)
    inertia.append(model.inertia_)

In [135]:
# Create a dictionary with the data to plot the Elbow curve
# YOUR CODE HERE!

elbow_df = pd.DataFrame({
    'k': k,
    'inertia': inertia
})

# Create a DataFrame with the data to plot the Elbow curve
# YOUR CODE HERE!

In [136]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# YOUR CODE HERE!

elbow_df.hvplot.line(
    x = 'k',
    y = 'inertia',
    xlabel = 'k',
    ylabel = 'Inertia',
    xticks = k
)

:Curve   [k]   (inertia)

#### Answer the following question: What is the best value for k?
**Question:** What is the best value for `k`?

**Answer:** The best value for k is 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In this section, you will use the K-Means algorithm with the best value for `k` found in the previous section to cluster the cryptocurrencies according to the price changes of cryptocurrencies provided.

1. Initialize the K-Means model with four clusters using the best value for `k`. 

2. Fit the K-Means model using the original data.

3. Predict the clusters to group the cryptocurrencies using the original data. View the resulting array of cluster values.

4. Create a copy of the original data and add a new column with the predicted clusters.

5. Create a scatter plot using hvPlot by setting `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.

In [137]:
# Initialize the K-Means model using the best value for k
# YOUR CODE HERE!

model_4 = KMeans(n_clusters=4, random_state=1)

In [138]:
# Fit the K-Means model using the scaled data
# YOUR CODE HERE!

model_4.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=1)

In [139]:
# Predict the clusters to group the cryptocurrencies using the scaled data
# YOUR CODE HERE!

prediction_4 = model_4.predict(df_market_data_scaled)

# View the resulting array of cluster values.
# YOUR CODE HERE!

display(prediction_4)

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1],
      dtype=int32)

In [140]:
# Create a copy of the DataFrame
# YOUR CODE HERE!

df_market_data_scaled_predicted = df_market_data_scaled.copy()

In [141]:
# Add a new column to the DataFrame with the predicted clusters
# YOUR CODE HERE!

df_market_data_scaled_predicted['customer segment 4'] = prediction_4

# Display sample data
# YOUR CODE HERE!

display(df_market_data_scaled_predicted)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [142]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# YOUR CODE HERE!

df_market_data_scaled_predicted.hvplot.scatter(
    x = 'price_change_percentage_24h',
    y = 'price_change_percentage_7d',
    by = 'customer segment 4',
    hover_cols = 'coin_id'
)

:NdOverlay   [customer segment 4]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis

In this section, you will perform a principal component analysis (PCA) and reduce the features to three principal components.

1. Create a PCA model instance and set `n_components=3`.

2. Use the PCA model to reduce to three principal components. View the first five rows of the DataFrame. 

3. Retrieve the explained variance to determine how much information can be attributed to each principal component.

4. Answer the following question: What is the total explained variance of the three principal components?

5. Create a new DataFrame with the PCA data. Be sure to set the `coin_id` index from the original DataFrame as the index for the new DataFrame. Review the resulting DataFrame.

In [143]:
# Create a PCA model instance and set `n_components=3`.
# YOUR CODE HERE!

pca = PCA(n_components=3)

In [144]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
# YOUR CODE HERE!

market_pca = pca.fit_transform(df_market_data)

# View the first five rows of the DataFrame. 
# YOUR CODE HERE!

display(market_pca)

array([[-3.41800963e+02, -5.13667755e+01,  1.25254709e+01],
       [-2.49420466e+02,  2.41175478e+01, -1.42314660e+01],
       [-4.02614721e+02, -1.18710737e+02,  2.48383966e+01],
       [-4.06752437e+02, -7.94872863e+01,  1.56633057e+00],
       [-3.82429948e+02, -1.03431959e+02,  1.67530727e+01],
       [-2.89125020e+02,  1.22871700e+01,  3.41638484e+01],
       [ 2.81514081e+01,  1.54987995e+02, -7.31265058e+01],
       [-1.74519832e+02,  8.02434934e+01, -3.03928300e+01],
       [-4.06613342e+02, -9.17830288e+01,  5.01614430e+00],
       [-3.11219887e+02, -1.43285351e+02,  6.08308028e+00],
       [-4.37478017e+01, -8.80040267e-01, -2.30882836e+01],
       [-4.02864243e+02, -1.18838225e+02,  2.45963216e+01],
       [-4.14237977e+02, -1.01768963e+02, -8.05327381e+00],
       [-2.16818545e+02,  8.49382259e+00,  5.59783020e+01],
       [-3.06196962e+02, -1.19529409e+01,  4.57326296e+00],
       [-2.56458443e+02, -1.27902916e+02, -2.19830267e+01],
       [-2.55261261e+02, -1.24797933e+02

In [145]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
# YOUR CODE HERE!

display(pca.explained_variance_ratio_)
pca_results = list(pca.explained_variance_ratio_)
print(pca_results[0] + pca_results[1] + pca_results[2])

array([9.76037313e-01, 2.30282949e-02, 7.48308214e-04])

0.9998139161298989


#### Answer the following question: What is the total explained variance of the three principal components?

**Question:** What is the total explained variance of the three principal components?

**Answer:** # The total explained variance of the three principal components is 99.98%

In [146]:
# Create a new DataFrame with the PCA data.
# Note: The code for this step is provided for you

market_pca_df = pd.DataFrame(
    market_pca,
    columns=['PCA1', 'PCA2', 'PCA3']
)

# Creating a DataFrame with the PCA data
# YOUR CODE HERE!

# Copy the crypto names from the original data
# YOUR CODE HERE!

market_pca_df["coin_id"] = df_market_data.index

# Set the coinid column as index
# YOUR CODE HERE!

market_pca_df = market_pca_df.set_index('coin_id')

# Display sample data
# YOUR CODE HERE!

display(market_pca_df)

PCA1        PCA2       PCA3
coin_id                                                 
bitcoin               -341.800963  -51.366775  12.525471
ethereum              -249.420466   24.117548 -14.231466
tether                -402.614721 -118.710737  24.838397
ripple                -406.752437  -79.487286   1.566331
bitcoin-cash          -382.429948 -103.431959  16.753073
binancecoin           -289.125020   12.287170  34.163848
chainlink               28.151408  154.987995 -73.126506
cardano               -174.519832   80.243493 -30.392830
litecoin              -406.613342  -91.783029   5.016144
bitcoin-cash-sv       -311.219887 -143.285351   6.083080
crypto-com-chain       -43.747802   -0.880040 -23.088284
usd-coin              -402.864243 -118.838225  24.596322
eos                   -414.237977 -101.768963  -8.053274
monero                -216.818545    8.493823  55.978302
tron                  -306.196962  -11.952941   4.573263
tezos                 -256.458443 -127.902916 -21.983027
okb                   -255.261261 -124.797933  17.121792
stellar               -364.332796  -49.102272 -20.410861
cosmos                -268.219369   30.466301  -8.224769
cdai                  -403.023491 -119.107827  24.449027
neo                   -229.661892   -8.803589   5.686160
wrapped-bitcoin       -341.672172  -51.088157  12.046052
leo-token             -377.733744 -110.558329  16.228922
huobi-token           -366.548441  -87.255215  14.133475
nem                   -145.551343   37.211474  69.779081
binance-usd           -402.470760 -118.671238  24.689086
iota                  -378.208662  -34.321847 -25.265565
vechain               -128.539445   65.675378 -59.716008
zcash                 -286.959559  -31.013635  -7.990442
theta-token            532.124138  516.120826 -13.315019
dash                  -403.345732 -117.010568  -3.148999
ethereum-classic      -389.140428 -116.155069   0.891223
ethlend               7755.778366 -361.972320 -11.117604
maker                 -367.838480  -46.016102  -0.000045
havven                 431.490135  255.732568 -97.419812
omisego                 79.668292  346.529605   3.861393
celsius-degree-token  1993.820798  823.817226  97.948664
ontology              -398.571139  -71.939902 -29.341220
ftx-token             -224.737127 -114.795270  23.570526
true-usd              -402.271474 -118.900574  24.885715
digibyte               -82.125235  275.234662 -74.559616

---

### Find the Best Value for k Using the PCA Data

In this section, you will use the elbow method to find the best value for `k` using the PCA data.

1. Code the elbow method algorithm and use the PCA data to find the best value for `k`. Use a range from 1 to 11. 

2. Plot a line chart with all the inertia values computed with the different values of `k` to visually identify the optimal value for `k`.

3. Answer the following questions: What is the best value for k when using the PCA data? Does it differ from the best k value found using the original data?

In [147]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
# YOUR CODE HERE!

k = list(range(1, 11))

In [148]:
# Create an empy list to store the inertia values
# YOUR CODE HERE!

pca_inertia = []

In [149]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
# YOUR CODE HERE!

for i in k:
    model = KMeans(n_clusters=i, random_state=1)
    model.fit(market_pca_df)
    pca_inertia.append(model.inertia_)

In [150]:
# Create a dictionary with the data to plot the Elbow curve
# YOUR CODE HERE!

elbow_pca_df = pd.DataFrame({
    'k': k,
    'inertia': pca_inertia
})
# Create a DataFrame with the data to plot the Elbow curve
# YOUR CODE HERE!


In [151]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# YOUR CODE HERE!

elbow_pca_df.hvplot.line(
    x = 'k',
    y = 'inertia',
    xlabel = 'k',
    ylabel = 'Inertia',
    xticks = k
)

:Curve   [k]   (inertia)

#### Answer the following questions: What is the best value for k when using the PCA data? Does it differ from the best k value found using the original data?
* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Best value of k is 2


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes it does differ, it went from k = 4 to k = 2

---

### Cluster Cryptocurrencies with K-means Using the PCA Data

In this section, you will use the PCA data and the K-Means algorithm with the best value for `k` found in the previous section to cluster the cryptocurrencies according to the principal components.

1. Initialize the K-Means model with four clusters using the best value for `k`. 

2. Fit the K-Means model using the PCA data.

3. Predict the clusters to group the cryptocurrencies using the PCA data. View the resulting array of cluster values.

4. Add a new column to the DataFrame with the PCA data to store the predicted clusters.

5. Create a scatter plot using hvPlot by setting `x="PC1"` and `y="PC2"`. Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point.

In [152]:
# Initialize the K-Means model using the best value for k
# YOUR CODE HERE!

model = KMeans(n_clusters=2, random_state=1)

In [153]:
# Fit the K-Means model using the PCA data
# YOUR CODE HERE!

model.fit(market_pca_df)

KMeans(n_clusters=2, random_state=1)

In [154]:
# Predict the clusters to group the cryptocurrencies using the PCA data
# YOUR CODE HERE!

prediction = model.predict(market_pca_df)
# View the resulting array of cluster values.
# YOUR CODE HERE!

display(prediction)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int32)

In [155]:
# Create a copy of the DataFrame with the PCA data
# YOUR CODE HERE!

market_pca_df_predicted = market_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
# YOUR CODE HERE!

market_pca_df_predicted['customer segment 2'] = prediction

# Display sample data
# YOUR CODE HERE!

display(market_pca_df_predicted.head())

PCA1        PCA2       PCA3  customer segment 2
coin_id                                                            
bitcoin      -341.800963  -51.366775  12.525471                   0
ethereum     -249.420466   24.117548 -14.231466                   0
tether       -402.614721 -118.710737  24.838397                   0
ripple       -406.752437  -79.487286   1.566331                   0
bitcoin-cash -382.429948 -103.431959  16.753073                   0

In [156]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# YOUR CODE HERE!

market_pca_df_predicted.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by = 'customer segment 2',
    hover_cols= 'coin_id'
)

:NdOverlay   [customer segment 2]
   :Scatter   [PCA1]   (PCA2,coin_id)

---

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

1. Create a composite plot using hvPlot and the plus (`+`) operator to contrast the Elbow Curve that you created to find the best value for `k` with the original and the PCA data.

2. Create a composite plot using hvPlot and the plus (`+`) operator to contrast the cryptocurrencies clusters using the original and the PCA data.

3. Answer the following question: After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

> **Rewind:** Back in Lesson 3 of Module 6, you learned how to create composite plots. You can look at that lesson to review how to make these plots; also, you can check [the hvPlot documentation](https://holoviz.org/tutorial/Composing_Plots.html).

In [157]:
# Composite plot to contrast the Elbow curves
# YOUR CODE HERE!
elbow_df.hvplot.line(
    x = 'k',
    y = 'inertia',
    xlabel = 'k',
    ylabel = 'Inertia',
    xticks = k,
    ylim = (0,300)) + elbow_pca_df.hvplot.line(
    x = 'k',
    y = 'inertia',
    xlabel = 'k',
    ylabel = 'Inertia',
    xticks = k
)
#display(inertia)

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [159]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

df_market_data_scaled_predicted.hvplot.scatter(
    x = 'price_change_percentage_24h',
    y = 'price_change_percentage_7d',
    by = 'customer segment 4',
    hover_cols = 'coin_id'
) + market_pca_df_predicted.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by = 'customer segment 2',
    hover_cols= 'coin_id'
)

:Layout
   .NdOverlay.I  :NdOverlay   [customer segment 4]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [customer segment 2]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Decreasing dimensionality with PCA and using fewer clusters as a result of said analysis points out that there might not be as many distinct groups as we initially thought with the original dataset. The clusters from PCA analysis shows that there might only be 2 distinct groups as opposed to 4 as initially thought with the initial dataset.